docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,188 @@
+
+### Goals
+
+- Logically separate transformation of Cluster API resources to infrastructure definitions from reconciliation of that infrastructure in Azure
+- Determine a method to migrate CAPZ's individual Azure service interfaces one-by-one to use ASO
+- Determine how ASO will be installed on Cluster API management clusters
+
+### Non-Goals/Future Work
+
+- Entirely eliminate direct usage of any Azure Go SDK
+- Reconcile long-running Azure operations asynchronously by waiting for
+  Kubernetes watch events from changes to the `status` of ASO resources instead
+  of periodically checking if the operation is complete
+- Reconcile different Azure resource types for the same workload cluster in parallel
+- Allow users to interact with ASO resources directly, out-of-band with CAPZ
+- Adopt Azure Workload Identity to authenticate with Azure APIs
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Users of new Azure features
+
+As a CAPZ user I want to take advantage of the latest and greatest features in Azure only available in the newest Azure API versions in my clusters managed by CAPZ. This would allow me to adopt CAPZ for a wider range of use cases and continue evolving my CAPZ clusters alongside the Azure API.
+
+#### Story 2 - Developer experience
+
+As a CAPZ developer I would like to only need to interact with the Kubernetes API server using Custom Resource Definitions representing Azure resources rather than having to deal with the impedance of interacting with the Azure APIs. This would enable me to focus more on the core value of CAPZ, transforming Cluster API resource definitions to Azure resource definitions.
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
+  - Install in capz-controller-manager `func main()`
+- GET requests are not as straightforward in the ASO model because BYO resources
+  may not initially exist in ASO
+- Current service pattern in CAPZ works for Azure services that are not
+  represented in ASO (resource health)
+- Each release of ASO can expose several Azure API versions of each resource,
+  including GA and preview API versions.
+- Must ensure CAPZ only operates against the ASO resources it manages, not ones managed by the user.
+
+### Security Model
+
+ASO requires credentials to interact with Azure which are documented here: https://azure.github.io/azure-service-operator/introduction/authentication/. Specifically, ASO needs the values of `$AZURE_CLIENT_ID`, `$AZURE_CLIENT_SECRET`, `$AZURE_SUBSCRIPTION_ID`, and `$AZURE_TENANT_ID` for Service Principal authentication, which are already set in the capz-controller-manager container's environment variables. Passing them along as chart values when installing ASO with Helm doesn't introduce any significant additional security risks.
```

Comment:
I believe the security model is more complicated than simply using the service principal credentials from the CAPZ controller. For example, each cluster in CAPZ can have its own Identity used to mutate Azure resources. How would this be handled with ASO?

Reply:
See https://azure.github.io/azure-service-operator/introduction/multitenant-deployment/#single-operator-multitenancy-default

Use the `serviceoperator.azure.com/credential-from` annotation on all of the resources associated with the the cluster to use a specific credential (rather than the global one).

Reply:
@matthchr Does ASO require the global credentials to be set if every resource has the `serviceoperator.azure.com/credential-from` annotation set?

Reply:
^ Assuming the answer is "no," I've updated this section with how I think this could work.

Reply:
Today you must create the global secret and it must contain at least a key for `AZURE_SUBSCRIPTION_ID`. All of the rest of the fields are optional. Technically you can set `AZURE_SUBSCRIPITON_ID` to anything (we don't validate it).

I just filed https://github.com/Azure/azure-service-operator/issues/2725 -- we should be able to do that for the next release v2.0.0 (it's a small change and making sure we have a good error).

Reply:
Just looping back around that the above linked issue is resolved now and will be in 2.0.0. You can omit the global secret. The actual k8s secret must still be created (it can be empty).